nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/spi/ConfigProviderResolverImpl.java

```diff
@@ -129,31 +126,34 @@
 
     // Gives access to deployed applications
     @Inject
-    ApplicationRegistry applicationRegistry;
-    
+    private ApplicationRegistry applicationRegistry;
+
     // This injects the configuration from the domain.xml magically
     // and for the correct server configuation
     @Inject
     @Named(ServerEnvironment.DEFAULT_INSTANCE_NAME)
     @Optional // PAYARA-2255 make optional due to race condition writing a missing entry into domain.xml
-    MicroprofileConfigConfiguration configuration;
+    private MicroprofileConfigConfiguration configuration;
 
     // a config used at the server level when there is no application associated with the thread
     private Config serverLevelConfig;
 
+    /**
+     * Sets the global {@link ConfigProviderResolver#instance()} to this instance.
+     */
     public ConfigProviderResolverImpl() {
+        LOG.finest("ConfigProviderResolverImpl()");
+        // the setInstance is not synchronized, but instance() method body is.
+        // this will block possible concurrent access.
+        synchronized (ConfigProviderResolver.class) {
+            LOG.log(Level.CONFIG, "Setting global ConfigProviderResolver instance to {0}", this);
+            ConfigProviderResolver.setInstance(this);
```

Comment:
This is executed too early if you ask me. The current instance is set as value BEFORE the instance is completely setup as HK2 service. So MP config calls can get routed to this class before HK2 is finished processing, resulting in errors (NPE since the @Inject isn't processed?)

Reply:
I agree, I was a bit too agressive and this is an antipattern (providing this before finishing constructor). But ... if we don't have other possibility, isn't the @PostConstruct method same situation?
From what I tested, @Inject will not cause NPE here, because HK2 will block any injection until it finishes PostConstruct method (constructor already finished). Bigger problem is programmatic usage of Globals class.

Reply:
Semantics of postconstruct guarantees that fields are set. The situation we're dealing with here is that someone triggered mp config API before this service was initialised. Postconstruct is therefore almost proper time as it is considered initialized after the method exits

Injections of this class via globals is indeed wrong